workflow/labels: switch to a scheduled trigger

wolfgangwalther opened the pull request:
This came out of the after-merge discussion in #415259.

Instead of relying on `workflow_run` events, we now trigger the labeling workflow by schedule. This avoids all permission/secrets problems of running in the `pull_request_review` context - and also gets rid of the "waiting for approval to run workflows" button for new contributors that sometimes comes up right now.

Also, it's more efficient. Previously, the labeling workflow would run on *any* `pull_request_review` event, which happens for all comments, too. So quite a few runs.

This will cause a delay of up to 1 hour with the current settings until approval labels are set. Depending on how long the job normally runs we can adjust the frequency. The workflow is written in a way that will work no matter what the frequency ends up being, even when it's interrupted by transient GHA failures: It will always look at all PRs which were updated since the last time the workflow ran successfully.

We also add the ability to run it manually via UI. This is useful:
- When fixing bugs in the labels workflow to run it all the way back to where the bug was introduced.
- When introducing new labels, to get a head start for a reasonable number of PRs immediately.

Of course, the workflow is also still run directly after Eval itself. This is also the only case that the actions/labeler steps will run, since they depend on the `pull_request` context.

As a side-effect this also handles #416719 better. I wouldn't close it with that, because old PRs are not labeled correctly, yet. But at least the job should not fail anymore.

## Things done

Of course, the interesting question will be how this behaves at the scale of Nixpkgs, compared to my lonely fork.

- [x] Tested in my fork.
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
workflow/labels: switch to a scheduled trigger

Instead of relying on workflow_run events, we now trigger the labeling
workflow by schedule. This avoids all permission/secrets problems of
running in the pull_request_review context - and also gets rid of the
"waiting for approval to run workflows" button for new contributors that
sometimes comes up right now.

Also, it's more efficient. Previously, the labeling workflow would run
on *any* pull_request_review event, which happens for all comments, too.
So quite a few runs.

This will cause a delay of up to 1 hour with the current settings until
approval labels are set. Depending on how long the job normally runs we
can adjust the frequency. The workflow is written in a way that will
work no matter what the frequency ends up being, even when it's
interrupted by transient GHA failures: It will always look at all PRs
which were updated since the last time the workflow ran successfully.

We also add the ability to run it manually via UI. This is useful:
- When fixing bugs in the labeler workflow to run it all the way back to
where the bug was introduced.
- When introducing new labels, to get a head start for a reasonable
amount of PRs immediately.

Of course, the workflow is also still run directly after Eval itself.
This is also the only case that the actions/labeler steps will run,
since they depend on the `pull_request` context.

philiptaron commented on .github/workflows/labels.yml:
```diff
@@ -6,12 +6,16 @@
 
 on:
+  schedule:
+    - cron: '37 * * * *'
```
Explainer: <https://crontab.guru/#37_*_*_*_*>

I'm partial to minute [28](https://en.wikipedia.org/wiki/28_(number)) myself (it's perfect), but [37](https://en.wikipedia.org/wiki/37_(number)) will do in a pinch.

philiptaron commented on .github/workflows/labels.yml:
```diff
@@ -31,57 +35,107 @@ jobs:
+            const cutoff = new Date(await (async () => {
+              // Always run for Pull Request triggers, no cutoff since there will be a single
+              // response only anyway.
+              if (context.payload.pull_request?.number) return 0
```
`new Date(0)` is `Wed Dec 31 1969 16:00:00 GMT-0800 (Pacific Standard Time)` aka [the UNIX epoch](https://en.wikipedia.org/wiki/Unix_time).

philiptaron commented on .github/workflows/labels.yml:
```diff
@@ -31,57 +35,107 @@ jobs:
+
+              // Manually triggered via UI. Will fallthrough to the last option if the
+              // updatedWithin parameter is not given.
+              const updatedWithin = Number.parseInt(process.env.UPDATED_WITHIN, 10)
```
Fun fact: `Number.parseInt` looks as far as it can forward but then ignores what's left, so if the value is `"10 hours"` it still returns 10 🤪

MattSturgeon replied:
Since this counts as user-input, should there be a validation step?

```yaml
- name: Validate updatedWithin input
  if: inputs.updatedWithin
  run: |
    # TODO: assert that it is a number
```

Not super critical, since the user input is always coming from someone with `actions: write` permission (i.e. a committer)

philiptaron commented on .github/workflows/labels.yml:
```diff
@@ -6,12 +6,16 @@
+  workflow_dispatch:
+    inputs:
+      updatedWithin:
+        description: 'Updated within [h]'
```
`[h]` means hours here.

philiptaron commented on .github/workflows/labels.yml:
```diff
@@ -31,57 +35,107 @@ jobs:
+                // Instead of checking the boolean artifact.expired, we will give us a minute to
+                // actually download the artifact in the next step and avoid that race condition.
+                log('Artifact expires at', artifact.expires_at)
+                if (new Date(artifact.expires_at) < new Date().getTime() + 60 * 1000) return;
```
Isn't this a type error?

```suggestion
                if (new Date(artifact.expires_at) < new Date(new Date().getTime() + 60 * 1000)) return;
```

philiptaron reviewed (commented):
That's fewer changes than I suspected would be needed.

MattSturgeon reviewed (commented):
My JavaScript is feeling rusty and I haven't really internalized this workflow yet. But here's my initial feedback :slightly_smiling_face: 

> Depending on how long the job normally runs we can adjust the frequency.

I suspect that given how many PR events nixpkgs gets, more or less _any_ value will still be an improvement, even ridiculously small intervals like running every few minutes. Although in reality such short intervals would likely run into issues with concurrency\* and/or cancellation.

Depending on how long the workflow actually takes to run, maybe the interval should be some multiplier of a typical run length? E.g. 2x or 3x average run?

\*Speaking of concurrency, this makes me wonder if we should use a non-cancelling concurrency group for the scheduled runs, so that if we ever did get a scenario where a run is still going when the next one is scheduled, the next one will wait in the queue while the first continues.

wolfgangwalther committed:
workflow/labels: save an API request

With the previous commit we now have the `before` labels available
already, which allows some simplification.

wolfgangwalther reviewed (commented):
Thanks for the great, detailed feedback, so far!

philiptaron reviewed (approved):
Let's give it a go. To @MattSturgeon's point about running more frequently, I agree that's the likely next PR, and it's as simple as going from `37 * * * *` to `7,22,37,52 * * * *`.

MattSturgeon commented:
Nice!

Looks like [`downloadArtifact`](https://github.com/actions/toolkit/blob/main/packages/artifact/docs/generated/interfaces/ArtifactClient.md#downloadartifact) takes an optional second argument _`options`_, of type [`DownloadArtifactOptions`](https://github.com/actions/toolkit/blob/main/packages/artifact/docs/generated/interfaces/DownloadArtifactOptions.md) & [`FindOptions`](https://github.com/actions/toolkit/blob/main/packages/artifact/docs/generated/interfaces/FindOptions.md).

`DownloadArtifactOptions` has a [`path`](https://github.com/actions/toolkit/blob/main/packages/artifact/docs/generated/interfaces/DownloadArtifactOptions.md#path) property:

> ### path
> • `Optional` **path**: `string`
> Denotes where the artifact will be downloaded to. If not specified then the artifact is download to GITHUB_WORKSPACE

We should be able to use this to specify a unique name.
